velox/functions/prestosql/aggregates/ReservoirSampleAccumulator.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <random>
+#include "velox/common/base/BitUtil.h"
+#include "velox/functions/lib/aggregates/ValueList.h"
+#include "velox/type/FloatingPointUtil.h"
+#include "velox/type/StringView.h"
+#include "velox/vector/DecodedVector.h"
+
+namespace facebook::velox::aggregate::prestosql {
+
+struct ReservoirSampleAccumulator {
+  ValueList samples;
+  ValueList initialSamples;
+  int64_t initialSeenCount = 0;
+  int64_t processedCount = 0;
+  int32_t maxSampleSize = -1;
+
+  explicit ReservoirSampleAccumulator() {}
+
+  bool isInitialized() const {
+    return maxSampleSize >= 0;
+  }
+
+  void addValueToReservoir(
+      const DecodedVector& decodedVector,
+      vector_size_t row,
+      HashStringAllocator* allocator) {
+    if (samples.size() < maxSampleSize) {
+      // LOG(INFO) << "[46] value : " << decodedVector.valueAt<>(row);
+      samples.appendValue(decodedVector, row, allocator);
+      return;
+    }
+
+    int64_t replaceIndex = -1;
+    if (!shouldReplaceSample(decodedVector, row, replaceIndex)) {
+      return;
+    }
+
+    replaceSampleAt(decodedVector, row, replaceIndex, allocator);
+  }
+
+  bool shouldReplaceSample(
+      const DecodedVector& decodedVector,
+      vector_size_t row,
+      int64_t& replaceIndex) {
+    uint64_t valueHash = computeValueHash(decodedVector, row);
+    uint64_t seed =
+        bits::hashMix(valueHash, static_cast<uint64_t>(processedCount));
+
+    std::mt19937_64 rng(seed);
```

Comment:
I am curious do you have to create a new seed everytime you need to check if a sample should be replaced ? If not rng /dist can be method variables and should make this much faster.